File: Problems/No.37_Sudoku_Solver/sudoku_solver.cpp
class Solution {
public:
    
    // check if the num in valid in this position
    bool safe(vector<vector<char>> &board, int row, int col, char c) {
        
        // parameters
        int block_row = (row/3)*3;
        int block_col = (col/3)*3;
        
        // start checking
        for (int i=0; i<9; i++) if (board[i][col]==c or board[row][i]==c) return false;
        for (int i=0; i<3; i++)
            for (int j=0; j<3; j++) 
                if (board[block_row+i][block_col+j] == c) return false;
        
        return true;
        
    }
    
    // sudoku solver using back propagating
    bool SudokuSolver(vector<vector<char>> &board) {
        
        // parameter
        bool check_if_solved;

        // start compute
        for (int row=0; row<9; row++) {
            for (int col=0; col<9; col++) {

                if (board[row][col]=='.') {

                    // start check with each valid num
                    for (char c='1'; c<='9'; c++) {
                        
                        if (safe(board,row,col,c)) {
                            
                            board[row][col]=c;
                            check_if_solved = SudokuSolver(board);

                            // verify if the sudoku is solved
                            if (check_if_solved) return true;
                            else board[row][col]='.';
                            
                        }
                        
                    }
                    
                    return false;
                    
                }
                
            }
        }
        
        return true;
        
    }
    
    void solveSudoku(vector<vector<char>>& board) {
        
        SudokuSolver(board);
        
    }
    
};

File: Problems/No.37_Sudoku_Solver/sudoku_solver.py
class Solution(object):
    def solveSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: None Do not return anything, modify board in-place instead.
        """
        
        # check if the num in valid in this position
        def safe(board, row, col, c):
            
            # parameters
            block_row = row - row%3
            block_col = col - col%3
            
            # start checking
            for i in range(9):
                if (board[row][i]==c or board[i][col]==c):
                    return False
            for i in range(3):
                for j in range(3):
                    if (board[block_row+i][block_col+j] == c):
                        return False
            
            return True
    
        # sudoku solver using back propagating
        def SudokuSolver(board):
            
            # parameters
            nums = ['1', '2', '3', '4', '5', '6', '7', '8', '9']
            check_if_ans = True
            
            # start compute
            for row in range(9):
                for col in range(9):
                    if (board[row][col] == '.'):
                        
                        # start check with each valid num
                        for c in nums:
                            if (safe(board, row, col, c)):
                                board[row][col] = c
                                check_if_ans = SudokuSolver(board)
                                
                                # verify if the sudoku is solved
                                if (check_if_ans):
                                    return True
                                else:
                                    board[row][col] = '.'
                        return False
            return True
        
        SudokuSolver(board)
